pkg/roachpb/metadata.proto

```diff
@@ -55,6 +70,9 @@ message ReplicaDescriptor {
   // higher replica_id.
   optional int32 replica_id = 3 [(gogoproto.nullable) = false,
       (gogoproto.customname) = "ReplicaID", (gogoproto.casttype) = "ReplicaID"];
+
+  // Type indicates which raft activities a replica participates in.
+  optional ReplicaType type = 4 [(gogoproto.nullable) = false];
```

Comment:
Hmm, wait a minute. This is non-nullable, which means it'll always be encoded on the wire. This means that the bytes before this PR and after this PR change even if learners aren't used. I think this is problematic because we use `RangeDescriptor` in `CPut` (for splits and merges). Unfortunately, I think this means that this needs to be nullable.

As a basic example, if you create a cluster with 19.1 and then switch to this binary, it shouldn't be able to carry out any splits -- the splits should fail on the CPut. Incidentally that's what @jordanlewis just saw over in https://github.com/cockroachdb/cockroach/issues/38183, maybe he ran into this?

Reply:
Assuming I'm not missing something here and things are peachy, two things should fall out of the fix:

1. ideally a unit test that can just verify this CPut-compatibility of protos that we know need it - basically just hard-code the marshaled bytes of these protos' zero values in the test and fail if they change
2. we have version upgrade tests, those ought to catch this. I think they may not because there aren't any splits done under the new version (and any queue-based splits failing doesn't fail the test). So we may just need to add an `ALTER TABLE SPLIT AT` and that should either fail or hang (both we can detect).

Reply:
Ooof those allocations are pretty unfortunate, no? Isn't this used in a good number of places?

I was surprised gogo was even serializing this field when set to the default value, but sure enough: https://github.com/cockroachdb/cockroach/blob/master/pkg/roachpb/metadata.pb.go#L952-L970. Removing the nullable = false annotation does in fact do the right thing:

```
	if m.Type != nil {
		dAtA[i] = 0x20
		i++
		i = encodeVarintMetadata(dAtA, i, uint64(*m.Type))
	}
```

I tried putting the same enum and proto into a proto3 syntax file and that also does the right thing. What are the issues embedding a proto3 struct into a proto2 one again?

```
func (m *ReplicaDescriptor3) MarshalTo(dAtA []byte) (int, error) {
	var i int
	_ = i
	var l int
	_ = l
	if m.NodeID != 0 {
		dAtA[i] = 0x8
		i++
		i = encodeVarintMetadata3(dAtA, i, uint64(m.NodeID))
	}
	if m.StoreID != 0 {
		dAtA[i] = 0x10
		i++
		i = encodeVarintMetadata3(dAtA, i, uint64(m.StoreID))
	}
	if m.ReplicaID != 0 {
		dAtA[i] = 0x18
		i++
		i = encodeVarintMetadata3(dAtA, i, uint64(m.ReplicaID))
	}
	if m.Type != 0 {
		dAtA[i] = 0x20
		i++
		i = encodeVarintMetadata3(dAtA, i, uint64(m.Type))
	}
	return i, nil
}
```

Finally, I wonder if the real issue here is the CPut (I guess I take back my "Yay for CPut" from the other PR). The proto spec is pretty loud about not relying on the exact bytes used to serialize a message and IIRC this is not the first time someone has hit this issue. What's the downside to turning the CPut into a Get and a Put and doing the not modified check on the deserialized structs? A bit of contention, but anything else? That seems to me to be the best long term fix to this

Reply:
> That seems to me to be the best long term fix to this

I'm not aware that there is any downside to this (and I agree with the upside). Wanna lead the charge? Probably want to coordinate with @jeffrey-xiao who introduced a similar problem (see #38020).